File: demo/src/main/java/com/link/AddTwoNumbers.java
package com.link;

public class AddTwoNumbers {
    public static void main(String[] args) {
        System.out.println(25 % 10);
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
//        ListNode head = new ListNode();
//        ListNode cur = head;
        ListNode head = null ,tail = null;
        // 用于两个节点的进位
        int k = 0;
        while (l1 != null || l2 != null) {
            int a1 = l1 != null ? l1.val : 0;
            int a2 = l2 != null ? l2.val : 0;
            int sum = a1 + a2 + k;
            if (head == null) {
                head = tail = new ListNode(sum % 10);
            } else {
                tail.next = new ListNode(sum % 10);
                tail = tail.next;
            }

//            cur.next = new ListNode(result % 10);
//            cur = cur.next;
            k = sum / 10;
            if (l1.next != null){
                l1 = l1.next;
            }
            if (l2.next != null){
                l2 = l2.next;
            }

        }
        // 这里是处理最后进位的
        if (k != 0){
            tail.next = new ListNode(k);
        }

        return head.next;
    }
}

File: godemo/com.link.main/AddTwoNumbers.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {

}
func addTwoNumbers(l1, l2 *ListNode) (head *ListNode) {
	var tail *ListNode
	var cur int
	for l1 != nil || l2 != nil {
		n1, n2 := 0, 0
		if l1 != nil {
			n1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			n2 = l2.Val
			l2 = l2.Next
		}
		sum := n1 + n2 + cur
		sum, cur = sum%10, sum/10
		if head == nil {
			head = &ListNode{Val: sum}
			tail = head
		} else {
			tail.Next = &ListNode{Val: sum}
			tail = tail.Next
		}
		if cur != 0 {
			tail.Next = &ListNode{Val: cur}
		}
	}
	return
}
